pennylane/gradients/spsa_gradient.py

```diff
@@ -0,0 +1,573 @@
+
+        >>> with qml.tape.QuantumTape() as tape:
+        ...     qml.RX(params[0], wires=0)
+        ...     qml.RY(params[1], wires=0)
+        ...     qml.RX(params[2], wires=0)
+        ...     qml.expval(qml.PauliZ(0))
+        ...     qml.var(qml.PauliZ(0))
+        >>> gradient_tapes, fn = qml.gradients.spsa_grad(tape)
+        >>> gradient_tapes
+        [<QuantumTape: wires=[0], params=3>, <QuantumTape: wires=[0], params=3>]
+
+        This can be useful if the underlying circuits representing the gradient
+        computation need to be analyzed. Here we see that for ``num_directions=1``,
+        the default, we obtain two tapes.
+
+        The output tapes can then be evaluated and post-processed to retrieve
+        the gradient:
+
+        >>> dev = qml.device("default.qubit", wires=2)
+        >>> fn(qml.execute(gradient_tapes, dev, None))
+        ((array(-0.58222637), array(0.58222637), array(-0.58222637)),
+         (array(1.05046797), array(-1.05046797), array(1.05046797)))
+
+
+        Devices that have a shot vector defined can also be used for execution, provided
+        the ``shots`` argument was passed to the transform:
+
+        >>> shots = (10, 100, 1000)
+        >>> dev = qml.device("default.qubit", wires=2, shots=shots)
+        >>> @qml.qnode(dev)
+        ... def circuit(params):
+        ...     qml.RX(params[0], wires=0)
+        ...     qml.RY(params[1], wires=0)
+        ...     qml.RX(params[2], wires=0)
+        ...     return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(0))
+        >>> params = np.array([0.1, 0.2, 0.3], requires_grad=True)
+        >>> qml.gradients.spsa_grad(circuit, shots=shots, h=1e-2)(params)
+        (((array(0.), array(0.), array(0.)), (array(0.), array(0.), array(0.))),
+         ((array(-1.4), array(-1.4), array(-1.4)),
+          (array(2.548), array(2.548), array(2.548))),
+         ((array(-1.06), array(-1.06), array(-1.06)),
+          (array(1.90588), array(1.90588), array(1.90588))))
+
+        The outermost tuple contains results corresponding to each element of the shot vector,
+        as is also visible by the increasing precision.
+        Note that the stochastic approximation and the fluctuations from the shot noise
+        of the device accumulate, leading to a very coarse-grained estimate for the gradient.
+    """
+    if argnum is None and not tape.trainable_params:
+        return _no_trainable_grad_new(tape, shots)
+
+    if validate_params:
+        if "grad_method" not in tape._par_info[0]:
+            gradient_analysis(tape, grad_fn=_spsa_grad_new)
+        diff_methods = grad_method_validation("numeric", tape)
+    else:
+        diff_methods = ["F" for i in tape.trainable_params]
+
+    if all(g == "0" for g in diff_methods):
+        return _all_zero_grad_new(tape, shots)
+
+    gradient_tapes = []
+    extract_r0 = False
+
+    coeffs, shifts = finite_diff_coeffs(n=n, approx_order=approx_order, strategy=strategy)
+
+    if 0 in shifts:
```

Comment:
You're right, the assumption is that `0` is the first shift in `shifts`. This is incorporated in the processing logic of parameter-shift rules as well as in `finite_diff_coeffs` and lets us skip any sorting steps after creating the differentiation recipes, and avoids indexing errors. If there is a place where additional documentation about this would be helpful, I'm happy for any suggestions you might have :)

Reply:
> and avoids indexing errors

you mean to say that `shifts` will sometimes be an empty list, so this ensures we don't panic in that case?

Reply:
Ah, no. I meant that if `0` was in `shifts`, but not first place, there would be quite a few functions in the `gradients` module that would have to find the index of the `0`, extract the corresponding coefficient correctly and so on... And that's a bit more error-prone I think.
Typically `shifts` is not empty, but if `0` is not in it (which happens rather often, actually), we don't have to extract an unshifted coefficient from the `coeffs`.